mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl

```diff
@@ -0,0 +1,343 @@
+	if d.Get("force_destroy") != nil {
+		if err := d.Set("force_destroy", d.Get("force_destroy")); err != nil {
+			return fmt.Errorf("Error updating force_destroy: %s", err)
+		}
+	}
+
+	// all other fields are immutable, don't do anything else
+	return nil
+}
+
+func resourceStorageFolderDelete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*transport_tpg.Config)
+	userAgent, err := tpgresource.GenerateUserAgentString(d, config.UserAgent)
+	if err != nil {
+		return err
+	}
+
+	bucket := d.Get("bucket").(string)
+	name := d.Get("name").(string)
+
+	var listError, deleteObjectError error
+	for deleteObjectError == nil {
+		res, err := config.NewStorageClient(userAgent).Objects.List(bucket).Prefix(name).Do()
+		if err != nil {
+			log.Printf("Error listing contents of bucket %s: %v", bucket, err)
+			// If we can't list the contents, try deleting the bucket anyway in case it's empty
+			listError = err
+			break
+		}
+
+		if len(res.Items) == 0 {
+			break // 0 items, folder empty
+		}
+
+		if !d.Get("force_destroy").(bool) {
+			deleteErr := fmt.Errorf("Error trying to delete bucket %s containing objects without force_destroy set to true", bucket)
+			log.Printf("Error! %s : %s\n\n", bucket, deleteErr)
+			return deleteErr
+		}
+		// GCS requires that a bucket be empty (have no objects or object
+		// versions) before it can be deleted.
+		log.Printf("[DEBUG] GCS Folder attempting to forceDestroy\n\n")
+
+		// Create a workerpool for parallel deletion of resources. In the
+		// future, it would be great to expose Terraform's global parallelism
+		// flag here, but that's currently reserved for core use. Testing
+		// shows that NumCPUs-1 is the most performant on average networks.
+		//
+		// The challenge with making this user-configurable is that the
+		// configuration would reside in the Terraform configuration file,
+		// decreasing its portability. Ideally we'd want this to connect to
+		// Terraform's top-level -parallelism flag, but that's not plumbed nor
+		// is it scheduled to be plumbed to individual providers.
+		wp := workerpool.New(runtime.NumCPU() - 1)
+
+		for _, object := range res.Items {
+			log.Printf("[DEBUG] Found %s", object.Name)
+			object := object
+
+			wp.Submit(func() {
+				log.Printf("[TRACE] Attempting to delete %s", object.Name)
+				if err := config.NewStorageClient(userAgent).Objects.Delete(bucket, object.Name).Generation(object.Generation).Do(); err != nil {
+					deleteObjectError = err
+					log.Printf("[ERR] Failed to delete storage object %s: %s", object.Name, err)
+				} else {
+					log.Printf("[TRACE] Successfully deleted %s", object.Name)
+				}
+			})
+		}
+
+		// Wait for everything to finish.
+		wp.StopWait()
+	}
+
+	if listError != nil {
+		return fmt.Errorf("could not delete non-empty folder due to error when listing contents: %v", listError)
+	}
+	log.Printf("[DEBUG] force_destroy value: %#v", d.Get("force_destroy").(bool))
+	if !d.Get("force_destroy").(bool) {
```

Comment:
If the folder with `force_destroy=false` that doesn't contain any sub-folder then it should get deleted without this error being thrown. 
We should do this validation after getting list of sub-folders. This was not caught in tests because there is no test for `force_destroy=false`. I recommend adding one test for the false value. 
Also It will same deletion logic for two different resources(objects and folders) so it would be good to write common logic for both and utilize that everywhere.

Reply:
updated the logic, if no subfolders are present and force_destroy set to false, allows to delete folder